src/utils/validator.ts

```diff
@@ -0,0 +1,80 @@
+import * as yup from 'yup';
+
+// import { IQueryParameters } from '../types/query-request';
+import { granulatiry, periodData } from './period';
+
+const periods = Object.keys(periodData);
+periods.push('all');
+
+export const validateQueryWithGroupData = yup.object({
+  period: yup.string().required('Missing period parameter').oneOf(periods),
+  func: yup.string().required('Missing func parameter'),
+  col: yup.string().required('Missing col parameter'),
+  granularity: yup.string().oneOf(granulatiry).notRequired(),
+  from: yup.number(),
+  to: yup.number(),
+});
+
+export const validateQuerySchema = yup.object({
+  period: yup.string().oneOf(periods),
+  func: yup.string(),
+  col: yup.string(),
+  limit: yup.number(),
+  offset: yup.number(),
+  granulatiry: yup.string().oneOf(granulatiry).notRequired(),
+});
+
+export const validateParams = yup.object({
+  id: yup.string().required('ID is required'),
+});
+
+// eslint-disable-next-line
+export function queryWithSortSchema(fields: string[]): yup.SchemaOf<any> {
+  return yup.object({
+    period: yup.string().oneOf(periods),
+    limit: yup.number().min(0).max(100),
+    offset: yup.number(),
+    sortBy: yup.string().oneOf([...fields]),
+    sortDirection: yup.string().oneOf(['DESC', 'ASC']).notRequired(),
+  });
+}
```

Comment:
Please change to the following:

```ts
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function queryWithSortSchema<Fields extends readonly string[]>(fields: Fields) {
  return yup.object({
    period: yup.mixed<TPeriod>().oneOf(periods as TPeriod[]),
    limit: yup.number().min(0).max(100),
    offset: yup.number(),
    sortBy: yup.mixed<Fields[number]>().oneOf([...fields]),
    sortDirection: yup.mixed<'DESC' | 'ASC'>().oneOf(['DESC', 'ASC']).notRequired(),
  });
}
```

And then in controller need to add `as const` for array of sortBy:
```ts
    const sortByFields = [
      'id',
      'timestamp',
      'difficulty',
      'size',
      'transactionCount',
    ] as const; // as const here
    try {
      const { sortBy, limit, offset, sortDirection, period } =
        queryWithSortSchema(sortByFields).validateSync(req.query);
     ....
```

And now typescript can work with it. Unfortunately, I didn't figure out how to define result type for the function, so leaved // eslint-disable comment

Reply:
@westalert let me know if you have a problem with this change

Reply:
![image](https://user-images.githubusercontent.com/69632363/127425140-ead00ceb-86a9-4caf-b153-077eaf48c31a.png)
![image](https://user-images.githubusercontent.com/69632363/127425150-e657e251-4bf3-4be5-a072-ec7e60bf5ae8.png)
I moved the fields in to the constants file.
and create the Fields types